### ISSUE TITLE
batchnorm 参数的问题

### ISSUE THREAD

**Issue by RayLiu2015** [PaddlePaddle/Paddle#12214]
![image](https://user-images.githubusercontent.com/13360993/42869437-9bcaf78a-8aa7-11e8-8ca2-85faa03bbcbc.png)

想请问下, batchnorm 参数中的 mean 和 savedMean 还有 variance 和 savedvariance 是什么关系, 我进行预测的时候应该用哪个?  我之前用的 variance, 但在一个模型中发现 variance 中的数字全都是 nan, 哪位同学能帮我解答一下呀

**Comment by kuke** [comment]
https://github.com/PaddlePaddle/Paddle/blob/4a0761781c10d2409f136b541375cbecbab651a6/paddle/fluid/operators/batch_norm_op.cc#L132-L134

这里有比较详细的参数解释，预测的时候只会用到mean和variance。另variance为nan的问题应该已经修复了，你可以用最新的代码重新训练一下模型

**Comment by qingqing01** [comment]
关于BatchNorm的moving variance出现NaN的情况，可以参考 https://github.com/PaddlePaddle/Paddle/issues/11445   这个Issue是因为在BatchNorm时，norm的维度上出现了只有1个元素，出现了NaN。

之前我们也遇到过这个问题，是在某次pass的最后一个batch的数据，只有一条数据，BN输入是conv层，feature的shape是[1, C, 1, 1]，这种情况，就会出现NaN。

你可以查下BN输入变量的shape，是否有问题，如果是这个问题，可以在feed数据时过滤下。

**Comment by RayLiu2015** [comment]
我的这个 batchnorm 输入维度是  [1, 128, 1, 1], 但如果我过滤了, 那使用什么数替代这些数呢?

**Comment by qingqing01** [comment]
@codeWorm2015  不太了解你的训练任务，我们在物体检测模型MobileNet-SSD的训练中，如果只有一条样本的，就跳过了，因为数据会shuffle，跳过这些数据不影响训练。 参考:

https://github.com/PaddlePaddle/models/blob/develop/fluid/object_detection/train.py#L158

另外，较新develop分支，BatchNorm Op的代码里处理了这种情况。

**Comment by kuke** [comment]
Closed due to be resolved

**Comment by lucywsq** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题， 我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持!
